revert #841

rmorshea opened the pull request:
closes: #920

and reverts changes from #841 as per discussion in #916

but preserve ability to declare attributes with snake_case

## Checklist

Please update this checklist as you complete each item:

- [x] Tests have been included for all bug fixes or added functionality.
- [x] The `changelog.rst` has been updated with any significant changes.
- [x] GitHub Issues which may be closed by this Pull Request have been linked.

Archmonger commented on src/idom/core/vdom.py:
```diff
@@ -133,11 +130,12 @@ def is_vdom(value: Any) -> bool:
+    import_source: ImportSourceDict | None = None,
 ) -> VdomDict:
-    """A helper function for creating VDOM elements.
+    """A helper function for creating VDOM dictionaries.
```
Probably want to keep this docstring change

Archmonger commented on scripts/fix_vdom_constructor_usage.py:
```diff
@@ -1,370 +0,0 @@
-from __future__ import annotations
```
Might want to keep this file, depending on whether we want to auto migrate `key=...`

rmorshea replied:
I think it'd probably be simpler to just have a normal deprecation warning for that. I can always go back in the history to find this as well.

rmorshea replied:
I'm actually going to handle moving key to be an attribute. There's a lot of places in the docs I don't want to fix so I may actually write a script to auto-update those things.

Archmonger commented on src/idom/core/vdom.py:
```diff
@@ -251,20 +233,44 @@ def constructor(
         elif (
-            # isinstance check on protocols is slow - use function attr pre-check as a
-            # quick filter before actually performing slow EventHandlerType type check
+            # isinstance check on protocols is slow, function attr check is a quick filter
```
might want to revert this comment change

Archmonger commented on src/idom/html.py:
```diff
@@ -392,20 +406,29 @@ def script(*children: str, key: Key | None = None, **attributes: Any) -> VdomDic
     """
+    model: VdomDict = {"tagName": "script"}
+
+    attributes, children = coalesce_attributes_and_children(attributes_and_children)
```
We might want to consider renaming `coalesce` to `split`?

Archmonger commented on src/idom/utils.py:
```diff
@@ -265,6 +266,12 @@ def _generate_vdom_children(
     )
 
 
+def _hypen_to_camel_case(string: str) -> str:
```
We probably don't want this behavior anymore.

It's possible some users want to use custom props such as `my-value=...` that contain hyphens. It's technically not a HTML5 violation. So ideally, we shouldn't touch hyphens.

Archmonger commented on tests/test_core/test_serve.py:
```diff
@@ -11,7 +11,7 @@
 
 
-EVENT_NAME = "on_event"
+EVENT_NAME = "onEvent"
```
Change needs confirmation

Archmonger commented on .github/workflows/test.yml:
```diff
@@ -23,7 +23,7 @@ jobs:
       session-arguments: --maxfail=3 --no-cov
       runs-on-array: '["ubuntu-latest", "macos-latest", "windows-latest"]'
-      python-version-array: '["3.7", "3.8", "3.9", "3.10", "3.11"]'
+      python-version-array: '["3.7", "3.8", "3.9", "3.10"]'
```
Need to re-add 3.11 tests

Archmonger commented on .pre-commit-config.yaml:
```diff
@@ -1,14 +1,14 @@
-    rev: 22.6.0
+    rev: 23.1.0
     hooks:
       - id: black
   - repo: https://github.com/pycqa/isort
-    rev: 5.6.3
+    rev: 5.12.0
     hooks:
       - id: isort
         name: isort
   - repo: https://github.com/pre-commit/mirrors-prettier
-    rev: "v2.5.1"
+    rev: v2.7.1
     hooks:
```
un-revert these bumps

rmorshea replied:
These were manually added. Was having some trouble with pre-commit for some reason.

rmorshea committed:
revert 764a5a8c9e083d4cc8ef69becb212953ce16b324

- but preserve ability to declare attributes with snake_case
- move 'key' to be an attribute rather than keyword argument
  in constructor + make auto converter for this change based
  on the one used for the "new" `*args` and `**kwargs`
  syntax we are reverting.

rmorshea committed:
add key rewrite script

rmorshea committed:
changelog

rmorshea committed:
fix types

Archmonger commented on docs/examples.py:
```diff
@@ -122,7 +122,7 @@ def Wrapper():
         text, set_text = idom.hooks.use_state(print_buffer.getvalue())
         print_buffer.set_callback(set_text)
-        return idom.html.pre(text, class_name="printout") if text else idom.html.div()
+        return idom.html.pre({"class": "printout"}, text) if text else idom.html.div()
```
`"class"` needs a revert to `"class_name"`. Also confirm throughout the docs we aren't missing any of those situations.

rmorshea replied:
```suggestion
        return idom.html.pre({"class_name": "printout"}, text) if text else idom.html.div()
```

Archmonger replied:
This appears to still be unresolved?

Archmonger commented on docs/source/about/changelog.rst:
```diff
@@ -23,7 +23,15 @@ more info, see the :ref:`Contributor Guide <Creating a Changelog Entry>`.
+
+**Deprecated**
+
+- :pull:`919` - Declaration of keys via keywork arguments in standard elements. A script
+  has been added to automatically convert old usages where possible.
```
Do we want to expand this to components as well, so we can ultimately fix our component type hints?

rmorshea replied:
I'm not quite sure how doing this would fix component type hints. Also components have a completely different interface to standard elements - they're effectively a normal function call with the exception of the reserved `key` parameter.

Archmonger replied:
I recall us discovering that our decorator can preserve type hints the original function's `*args, *kwargs` if we aren't trying to add a new kwarg (`key=...`).

rmorshea replied:
That's true, though we'd need a wrapper object around keys in order to distinguish them from other positional arguments. We could implement that as shown in [this gist](https://gist.github.com/5938469562e51a8c4617a8f472c4c684), but we should definitely do that in a separate PR though. Perhaps an opinion from @numpde here would be useful too.

rmorshea replied:
This also opens up other questions. Should we allow or require keys for standard elements to be declared in a similar manner? That is, as an optional first positional argument instead of via the normal attributes dict?

Archmonger replied:
> as an optional first positional argument instead of via the normal attributes dict

I think you might have mistyped

EDIT: Oh I think I'm getting what you meant. As in, should the `Key` class be usable with `idom.html` as well.

I guess the idea is, if we're doing `Key(...)` just for components, then at that point the props dict on `idom.html` might feel weird 🙃

But they both have caveats... so I would say let's continue with the current `dict` for `idom.html` and `kwarg` for `@component` for now. If we think of a better way of doing it down the line, that'll be a different major version release.

Archmonger commented on docs/source/guides/adding-interactivity/components-with-state/_examples/isolated_state/main.py:
```diff
@@ -29,14 +29,14 @@ def handle_more_click(event):
             html.button(
-                f"{('Show' if show_more else 'Hide')} details",
-                on_click=handle_more_click,
+                {"onClick": handle_more_click},
```
Need to confirm throughout the docs we are using the new underscore syntax.

Probably need to update the script to handle this.

rmorshea replied:
Yup. I think we can do this in a separate PR. I guess we have one more before renaming.

rmorshea replied:
Granted using camelCase is completely valid.

Archmonger commented on src/idom/_console/rewrite_key_declarations.py:
```diff
@@ -22,7 +21,7 @@
 @click.command()
 @click.argument("paths", nargs=-1, type=click.Path(exists=True))
-def update_html_usages(paths: list[str]) -> None:
+def rewrite_key_declarations(paths: list[str]) -> None:
```
maybe just `rewrite_keys` instead?

If we want this to handle the underscore syntax for props, this might become `v1_syntax_update`

rmorshea committed:
apply rewrites

rmorshea commented:
@Archmonger I think we can go forward with the rename after this.

Archmonger commented on src/idom/core/hooks.py:
```diff
@@ -281,7 +277,7 @@ def __init__(
     def render(self) -> VdomDict:
         current_hook().set_context_provider(self)
-        return vdom("", *self.children)
+        return {"tagName": "", "children": self.children}
```
Is there a reason this was switch to a raw VDOM dict?

rmorshea replied:
Just faster. No particular reason.

Archmonger commented on src/idom/core/hooks.py:
```diff
@@ -144,15 +140,15 @@ def use_effect(
             async_function = cast(_AsyncEffectFunc, function)
 
-            def sync_function() -> Optional[_EffectCleanFunc]:
+            def sync_function() -> _EffectCleanFunc | None:
```
We might want to add `pyupgrade` as a precommit-ci hook to automagically handle these

rmorshea replied:
That would be great to have

Archmonger replied:
I have a pretty exhaustive "zero effort auto fixes" precommit-ci config in Conreq's App store branch, if you want to take a peek.

https://github.com/Archmonger/Conreq/blob/d3740f4b7a340491c886dd114b7f2f40c398c4a1/.pre-commit-config.yaml

rmorshea replied:
Feel free to make a PR for this. I don't have time to check this out at the moment.

Archmonger replied:
Sure, I'll PR after v1.

Archmonger commented on src/idom/html.py:
```diff
@@ -280,7 +280,18 @@
 def _(*children: Any, key: Key | None = None) -> VdomDict:
     """An HTML fragment - this element will not appear in the DOM"""
-    return vdom("", *children, key=key)
+    attributes, coalesced_children = separate_attributes_and_children(children)
```
Need to rename `coalesced_children` to `children`. We should do a global search on `coalesce` just to clean up references.

Archmonger commented on src/idom/html.py:
```diff
@@ -392,20 +406,29 @@ def script(*children: str, key: Key | None = None, **attributes: Any) -> VdomDic
     function that is called when the script element is removed from the tree, or when
```
Unrelated to this PR...
We should add a disclaimer here to warn users not to use raw user inputs (or any potentially untrusted data source) within the script contents to avoid `XSS` attacks.

rmorshea replied:
Can you create an issue for this?

Archmonger commented on src/idom/utils.py:
```diff
@@ -309,6 +308,9 @@ def _vdom_attr_to_html_str(key: str, value: Any) -> tuple[str, str]:
 # see list of HTML attributes with dashes in them:
 # https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes#attribute_list
-_DASHED_HTML_ATTRS = {"accept_charset", "http_equiv"}
+_DASHED_HTML_ATTRS = {"acceptCharset", "httpEquiv"}
```
tbh these globals are innocent enough to be public instead of private.

It's possible some external libraries might want to rely on it some day, despite it being undocumented.

Archmonger commented on tests/test_core/test_component.py:
```diff
@@ -35,19 +35,19 @@ def SimpleParamComponent(tag):
+        return idom.html.div(kwargs, args)
 
-    assert ComponentWithVarArgsAndKwargs("hello", "world", my_attr=1).render() == {
+    assert ComponentWithVarArgsAndKwargs("hello", "world", myAttr=1).render() == {
```
Is the camelCase intentional here?

rmorshea replied:
Nope, not sure why that was camel in the first place given this is mostly a reversion.

rmorshea committed:
add custom vdom constructor decorator

rmorshea committed:
change to snake

rmorshea committed:
fix err msg

rmorshea committed:
cast keys to strings

rmorshea committed:
make pub

rmorshea committed:
fix dashed html attrs

rmorshea committed:
rename to rewrite-keys

rmorshea committed:
fix types

rmorshea committed:
allow ints in vdom spec for keys

rmorshea committed:
fix types

Archmonger commented on docs/source/reference/_examples/simple_dashboard.py:
```diff
@@ -83,10 +83,10 @@ def update_value(value):
 
     return idom.html.fieldset(
-        idom.html.legend(label, style={"font_size": "medium"}),
+        {"class": "number-input-container"},
```
`class` -> `class_name`

rmorshea replied:
```suggestion
        {"class_name": "number-input-container"},
```

rmorshea committed:
camelCase to snake_case + rewrite util

rmorshea committed:
fix types

rmorshea committed:
get cov

rmorshea committed:
fix style

rmorshea commented:
I'd like to get this out, so I'm going to merge and do a pre-release

Archmonger commented:
I'll do a post-merge review tomorrow. If there's any issues we can handle it in a separate release.

Archmonger commented on src/idom/_console/rewrite_camel_case_props.py:
```diff
@@ -0,0 +1,88 @@
+
+@click.command()
+@click.argument("paths", nargs=-1, type=click.Path(exists=True))
+def rewrite_camel_case_props(paths: list[str]) -> None:
```
This has issues. Currently it does not preserve dashes `-`, so `data-success` gets turned into `data_success`.

Also it adds an empty `VdomAttributes` dict no reason, For example `html.hr()` becomes `html.hr({})`

rmorshea replied:
`data_success` will be handled properly. I think I did that explicitly to have uniform styling. I can change that though. The latter case is definitely an issue.

Archmonger replied:
Yeah, I think `data-` and `aria-` shouldn't be auto-converted. It's fine if IDOM's client can automatically interpret underscored `data_` attributes, but the suggested styling should remain as a dash since our API allows it.

Archmonger commented on src/idom/_console/rewrite_keys.py:
```diff
@@ -0,0 +1,111 @@
+
+@click.command()
+@click.argument("paths", nargs=-1, type=click.Path(exists=True))
+def rewrite_keys(paths: list[str]) -> None:
```
We probably want to combine both of these into a more general `v1_syntax_update` command. I think that name makes sense, since we're likely to delete these commands once we reach v2.

Archmonger replied:
not needed so I'm closing this one

Archmonger commented on docs/source/_exts/custom_autosectionlabel.py:
```diff
@@ -30,6 +30,7 @@ def get_node_depth(node: Node) -> int:
 
 def register_sections_as_label(app: Sphinx, document: Node) -> None:
     docname = app.env.docname
+    print(docname)
```
Unneeded print?

Archmonger commented on docs/source/_exts/custom_autosectionlabel.py:
```diff
@@ -66,7 +67,7 @@ def register_sections_as_label(app: Sphinx, document: Node) -> None:
 
 
-def setup(app: Sphinx) -> dict[str, Any]:
+def setup(app: Sphinx) -> Dict[str, Any]:
```
No need to use `typing.Dict` here.

Archmonger commented on docs/source/guides/adding-interactivity/dangers-of-mutability/_examples/moving_dot_broken.py:
```diff
@@ -14,25 +14,29 @@ def handle_pointer_move(event):
+            {
+                "style": {
+                    "position": "absolute",
+                    "backgroundColor": "red",
```
Should we convert styles to `snake_case`?

Archmonger commented on docs/source/guides/adding-interactivity/multiple-state-updates/_examples/set_color_3_times.py:
```diff
@@ -15,9 +15,11 @@ def handle_reset(event):
     return html.div(
         html.button(
-            "Set Color", on_click=handle_click, style={"background_color": color}
+            {"on_click": handle_click, "style": {"backgroundColor": color}}, "Set Color"
```
Tons of `backgroundColor` references in the docs. We might want them to be `background_color`.

Archmonger commented on tests/test_core/test_serve.py:
```diff
@@ -115,8 +115,8 @@ async def handle_event():
         return idom.html.div(
-            idom.html.button(on_click=block_forever),
-            idom.html.button(on_click=handle_event),
+            idom.html.button({"onClick": block_forever}),
```
`onClick` -> `on_click`
There's a few instances of unconverted values in the tests. This might indicate the conversion script isn't working in some edge cases.

Archmonger commented on docs/source/guides/adding-interactivity/state-as-a-snapshot/_examples/print_chat_message.py:
```diff
@@ -16,24 +16,28 @@ async def handle_submit(event):
     return html.form(
+        {"on_submit": handle_submit, "style": {"display": "inline-grid"}},
         html.label(
+            {},
```
There's a handful of these empty dicts in this PR.

I recommend doing a search through this diff page for `{}`

Archmonger commented:
Been testing this out in Conreq. I have a couple of comments

1) Using `custom_vdom_constructor` on `_` and `script` elements causes pylint errors 
   - We should probably handle this through overloads and helper functions instead.
   - ![image](https://user-images.githubusercontent.com/16909269/220802436-daa76234-bd5f-49c8-a185-b0e6d2242819.png)

2) On some of my `idom.html` elements, I was constructing the vdom dict using function calls (ex. `generate_props(val_1, val_2)`). The `rewrite-keys` script then would add another vdom dict above it, ultimately breaking things.
   - Should we update the script to run a type-check on the first arg of an `idom.html` call (and then combine them with the pipe `|` operator if needed)?
   - The following error probably needs more verbosity in order to better catch things like this:

```
  File "C:\Users\Markg\Documents\Repositories\Conreq\.venv\lib\site-packages\idom\core\layout.py", line 202, in _render_model
    new_state.model.current = {"tagName": raw_model["tagName"]}
KeyError: 'tagName'
```

rmorshea commented:
Unfortunately I don't think there's a good great way handle that case you're experiencing. That first posarg could also be a child rather than a props dict. I think we can just produce a more helpful error.

Archmonger commented:
Yeah no problem. To be honest it probably doesn't need the extra robustness, since as far as I can tell I'm the only one "stress testing" IDOM.

The error improvement would be a good addition though
